File: libs/ontology-kernel/src/models/object_type.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct ObjectType {
    pub id: Uuid,
    pub name: String,
    pub display_name: String,
    pub description: String,
    pub primary_key_property: Option<String>,
    pub icon: Option<String>,
    pub color: Option<String>,
    pub owner_id: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct CreateObjectTypeRequest {
    pub name: String,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub primary_key_property: Option<String>,
    pub icon: Option<String>,
    pub color: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateObjectTypeRequest {
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub primary_key_property: Option<String>,
    pub icon: Option<String>,
    pub color: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ListObjectTypesQuery {
    pub page: Option<i64>,
    pub per_page: Option<i64>,
    pub search: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ListObjectTypesResponse {
    pub data: Vec<ObjectType>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
}

File: openfoundry-go/libs/ontology-kernel/models/object_type.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ObjectType mirrors `libs/ontology-kernel/src/models/object_type.rs`
// `struct ObjectType` 1:1.
type ObjectType struct {
	ID                  uuid.UUID `json:"id"                    db:"id"`
	Name                string    `json:"name"                  db:"name"`
	DisplayName         string    `json:"display_name"          db:"display_name"`
	Description         string    `json:"description"           db:"description"`
	PrimaryKeyProperty  *string   `json:"primary_key_property"  db:"primary_key_property"`
	Icon                *string   `json:"icon"                  db:"icon"`
	Color               *string   `json:"color"                 db:"color"`
	OwnerID             uuid.UUID `json:"owner_id"              db:"owner_id"`
	CreatedAt           time.Time `json:"created_at"            db:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"            db:"updated_at"`
}

// CreateObjectTypeRequest mirrors `struct CreateObjectTypeRequest`.
type CreateObjectTypeRequest struct {
	Name               string  `json:"name"`
	DisplayName        *string `json:"display_name,omitempty"`
	Description        *string `json:"description,omitempty"`
	PrimaryKeyProperty *string `json:"primary_key_property,omitempty"`
	Icon               *string `json:"icon,omitempty"`
	Color              *string `json:"color,omitempty"`
}

// UpdateObjectTypeRequest mirrors `struct UpdateObjectTypeRequest`.
type UpdateObjectTypeRequest struct {
	DisplayName        *string `json:"display_name,omitempty"`
	Description        *string `json:"description,omitempty"`
	PrimaryKeyProperty *string `json:"primary_key_property,omitempty"`
	Icon               *string `json:"icon,omitempty"`
	Color              *string `json:"color,omitempty"`
}

// ListObjectTypesQuery mirrors `struct ListObjectTypesQuery`.
type ListObjectTypesQuery struct {
	Page    *int64  `json:"page,omitempty"`
	PerPage *int64  `json:"per_page,omitempty"`
	Search  *string `json:"search,omitempty"`
}

// ListObjectTypesResponse mirrors `struct ListObjectTypesResponse`.
type ListObjectTypesResponse struct {
	Data    []ObjectType `json:"data"`
	Total   int64        `json:"total"`
	Page    int64        `json:"page"`
	PerPage int64        `json:"per_page"`
}
